domain/relation/service/relation.go

```diff
@@ -962,3 +983,90 @@ func (s *Service) SetRelationUnitSettings(
 	}
 	return s.st.SetRelationUnitSettings(ctx, relationUnitUUID, settings)
 }
+
+// ImportRelations sets relations imported in migration. It first builds all the
+// relations to insert from the arguments, then inserts them at the end so as to
+// wait as long as possible before turning into a write transaction.
+func (s *Service) ImportRelations(ctx context.Context, args relation.ImportRelationsArgs) error {
+	for _, arg := range args {
+		relUUID, err := s.importRelation(ctx, arg)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		for _, ep := range arg.Endpoints {
+			err = s.importRelationEndpoint(ctx, relUUID, ep)
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Service) importRelation(ctx context.Context, arg relation.ImportRelationArg) (corerelation.UUID, error) {
+	var relUUID corerelation.UUID
+
+	eps := arg.Key.EndpointIdentifiers()
+
+	switch len(eps) {
+	case 1:
+		// Peer relations are implicitly imported during migration of applications
+		// during the call to CreateApplication.
+		var err error
+		relUUID, err = s.st.GetPeerRelationUUIDByEndpointIdentifiers(ctx, eps[0])
+		if err != nil {
+			return relUUID, errors.Errorf("getting peer relation %d by endpoint %q: %w", arg.ID, eps[0], err)
+		}
+	case 2:
+		idep1, err := relation.NewCandidateEndpointIdentifier(eps[0].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[0].String(), err)
+		}
+		idep2, err := relation.NewCandidateEndpointIdentifier(eps[1].String())
+		if err != nil {
+			return relUUID, errors.Errorf("parsing endpoint identifier %q: %w", eps[1].String(), err)
+		}
+
+		relUUID, err = s.st.SetRelationWithID(ctx, idep1, idep2, uint64(arg.ID))
+		if err != nil {
+			return relUUID, errors.Capture(err)
+		}
+	default:
+		return relUUID, errors.Errorf("unexpected number of endpoints %d for %q", len(eps), arg.Key)
+	}
+	return relUUID, nil
+}
+
+func (s *Service) importRelationEndpoint(ctx context.Context, relUUID corerelation.UUID, ep relation.ImportEndpoint) error {
+	appID, err := s.st.GetApplicationIDByName(ctx, ep.ApplicationName)
+	if err != nil {
+		return err
+	}
+
+	err = s.st.SetRelationApplicationSettings(ctx, relUUID, appID, settingsMap(ep.ApplicationSettings))
+	if err != nil {
+		return err
+	}
+	for unitName, settings := range ep.UnitSettings {
+		err = s.st.EnterScope(ctx, relUUID, unit.Name(unitName), settingsMap(settings))
```

Comment:
We don't need to do the subordinate checks that come with enter scope, we could just have a function to set the relation_unit and relation_unit settings directly.

This could also be moved down a layer.

Reply:
leaving as is.